src/openrct2/Editor.cpp

```diff
@@ -399,57 +399,49 @@ namespace OpenRCT2::Editor
+        };
+
+        for (auto& pair : kBasicCheckPairs)
         {
-            if (!EditorCheckObjectGroupAtLeastOneSurfaceSelected(false))
-            {
-                return { ObjectType::FootpathSurface, STR_AT_LEAST_ONE_FOOTPATH_NON_QUEUE_SURFACE_OBJECT_MUST_BE_SELECTED };
-            }
-            if (!EditorCheckObjectGroupAtLeastOneSurfaceSelected(true))
+            if (!EditorCheckObjectGroupAtLeastOneSelected(pair.first))
             {
-                return { ObjectType::FootpathSurface, STR_AT_LEAST_ONE_FOOTPATH_QUEUE_SURFACE_OBJECT_MUST_BE_SELECTED };
-            }
-            if (!EditorCheckObjectGroupAtLeastOneSelected(ObjectType::FootpathRailings))
-            {
-                return { ObjectType::FootpathRailings, STR_AT_LEAST_ONE_FOOTPATH_RAILING_OBJECT_MUST_BE_SELECTED };
+                return { pair.first, pair.second };
             }
         }
 
-        if (!EditorCheckObjectGroupAtLeastOneSelected(ObjectType::Ride))
-        {
-            return { ObjectType::Ride, STR_AT_LEAST_ONE_RIDE_OBJECT_MUST_BE_SELECTED };
-        }
-        if (!EditorCheckObjectGroupAtLeastOneSelected(ObjectType::Station))
+        // Beyond this point, all checks don't apply to the track designer/manager.
+        const bool isTrackDesignerManager = gScreenFlags & (SCREEN_FLAGS_TRACK_DESIGNER | SCREEN_FLAGS_TRACK_MANAGER);
+        if (isTrackDesignerManager)
         {
-            return { ObjectType::Station, STR_AT_LEAST_ONE_STATION_OBJECT_MUST_BE_SELECTED };
+            return { ObjectType::None, STR_NONE };
         }
 
-        if (!EditorCheckObjectGroupAtLeastOneSelected(ObjectType::TerrainSurface))
+        if (!EditorCheckObjectGroupAtLeastOneSurfaceSelected(false))
         {
-            return { ObjectType::TerrainSurface, STR_AT_LEAST_ONE_TERRAIN_SURFACE_OBJECT_MUST_BE_SELECTED };
+            return { ObjectType::FootpathSurface, STR_AT_LEAST_ONE_FOOTPATH_NON_QUEUE_SURFACE_OBJECT_MUST_BE_SELECTED };
         }
-        if (!EditorCheckObjectGroupAtLeastOneSelected(ObjectType::TerrainEdge))
+        if (!EditorCheckObjectGroupAtLeastOneSurfaceSelected(true))
         {
-            return { ObjectType::TerrainEdge, STR_AT_LEAST_ONE_TERRAIN_EDGE_OBJECT_MUST_BE_SELECTED };
+            return { ObjectType::FootpathSurface, STR_AT_LEAST_ONE_FOOTPATH_QUEUE_SURFACE_OBJECT_MUST_BE_SELECTED };
         }
 
-        if (!isTrackDesignerManager)
-        {
-            if (!EditorCheckObjectGroupAtLeastOneSelected(ObjectType::ParkEntrance))
-            {
-                return { ObjectType::ParkEntrance, STR_PARK_ENTRANCE_TYPE_MUST_BE_SELECTED };
-            }
+        constexpr std::pair<ObjectType, StringId> kParkCheckPairs[] = {
+            { ObjectType::FootpathRailings, STR_AT_LEAST_ONE_FOOTPATH_RAILING_OBJECT_MUST_BE_SELECTED },
+            { ObjectType::ParkEntrance, STR_PARK_ENTRANCE_TYPE_MUST_BE_SELECTED },
+            { ObjectType::Water, STR_WATER_TYPE_MUST_BE_SELECTED },
```

Comment:
Do we really not need to have a "water" type selected in the track designer? Won’t it mess with the colours (considering it’s really a palette)?

Reply:
I believe that's exactly why the check is there -- because the game relies on fetching this 'water' palette in several places. The tab is hidden in the track designer/manager, anyway, but yeah, it seems the object should be selected.

Reply:
What I meant is: in your code, it is **not** checked if you’re in the track designer/manager, but _only_ in the scenario editor and in-game. That to me seems like an oversight.

Reply:
You're right, but that's not something I've changed in this PR. In the develop version of the function, the water object check is already inside an `if (!isTrackDesignerManager)` block, so indeed wouldn't be run in the track designer or manager. As I speculated yesterday, I think it might be a conscious choice, in that the user can't select a custom water object in those scenes.

Reply:
Maybe, but that doesn’t mean we have to stick with that. You’re changing the code around anyway, might as well fix this while you’re at it. It’s two lines.

Reply:
I'm not sure I agree. This PR deliberately tries to refactor the function _without_ changing the logic, after all. I'd you think this warrants a change, I would prefer you file a follow-up, honestly.

Reply:
For two lines? I don’t see how that makes sense.

Reply:
I would really rather not change any logic in this PR, especially for something that isn't even exposed through the UI. Asking for it to be changed now is what you'd call bikeshedding, imo.

Reply:
I agree with @AaronVanGeffen, lets stick to pure refactor and we can tackle the logical change afterwards.